File: loglib/src/main/java/com/robog/loglib/SLog.java
package com.robog.loglib;

import android.annotation.SuppressLint;
import android.content.Context;

import com.robog.loglib.database.LogBeanDao;
import com.robog.loglib.database.LogBeanDaoImpl;

import java.io.File;
import java.util.List;

import static com.robog.loglib.FileLog.FILE_PATH;
import static com.robog.loglib.Util.wrapLogBean;

/**
 * Created by yuxingdong on 2018/6/21.
 * <p>
 * 需要权限:
 * <li>{@link android.Manifest.permission#WRITE_EXTERNAL_STORAGE}
 * <li>{@link android.Manifest.permission#READ_EXTERNAL_STORAGE}
 * <p/>
 */
public final class SLog {

    public static final String TAG = "SLog";
    private static LogStrategy sLogStrategy;
    @SuppressLint("StaticFieldLeak")
    private static Context sContext;

    private static final int EXCEPTION = -1;
    private static final int VERBOSE = 2;
    private static final int DEBUG = 3;
    private static final int INFO = 4;
    private static final int WARN = 5;
    private static final int ERROR = 6;

    public static boolean sDebug;
    public static String sDeviceInfo;
    public static int sDbCheckInterval;
    public static int sDbThreshold;
    public static int sLogMode;
    public static int sCacheSize;

    public static void init(Context context) {
        init(context, new LogConfig.Builder().build());
    }

    public static void init(Context context, LogConfig logConfig) {
        sContext = context.getApplicationContext();

        sDebug = logConfig.debug();
        sDeviceInfo = logConfig.deviceInfo();
        sDbCheckInterval = logConfig.dbCheckInterval();
        sDbThreshold = logConfig.dbThreshold();
        sLogMode = logConfig.logMode();
        sCacheSize = logConfig.cacheSize();

        if (sLogMode == LogMode.DATABASE) {
            sLogStrategy = DatabaseLog.create(sContext);
        } else if (sLogMode == LogMode.FILE) {
            sLogStrategy = FileLog.create();
        }
    }

    public static void v(String tag, Exception e) {
        v(tag, e.toString());
    }

    public static void v(String msg) {
        v(TAG, msg);
    }

    public static void v(String tag, String msg) {
        if (sDebug) {
            SLog.v(tag, msg);
        }
        print(tag, msg, VERBOSE);
    }

    public static void d(String tag, Exception e) {
        d(tag, e.toString());
    }

    public static void d(String msg) {
        d(TAG, msg);
    }

    public static void d(String tag, String msg) {
        if (sDebug) {
            SLog.d(tag, msg);
        }
        print(tag, msg, DEBUG);
    }

    public static void i(String tag, Exception e) {
        i(tag, e.toString());
    }

    public static void i(String msg) {
        i(TAG, msg);
    }

    public static void i(String tag, String msg) {
        if (sDebug) {
            SLog.i(tag, msg);
        }
        print(tag, msg, INFO);
    }

    public static void w(String tag, Exception e) {
        w(tag, e.toString());
    }


    public static void w(String msg) {
        w(TAG, msg);
    }

    public static void w(String tag, String msg) {
        if (sDebug) {
            SLog.e(tag, msg);
        }
        print(tag, msg, WARN);
    }

    public static void e(String tag, Exception e) {
        e(tag, e.toString());
    }

    public static void e(String msg) {
        e(TAG, msg);
    }

    public static void e(String tag, String msg) {
        if (sDebug) {
            SLog.e(tag, msg);
        }
        print(tag, msg, ERROR);
    }

    public static int cacheSize() {
        return LogCache.CACHE.size();
    }

    /**
     * 将所有缓存数据推入数据库用于上传
     */
    public static void flush() {
        LogCache.get().flush();
    }

    /**
     * 获取所有数据
     */
    public static List<LogBean> getAllLog() {
        if (sContext == null) {
            throw new RuntimeException("Please init SLog first!");
        }
        final LogBeanDao logBeanDao = LogBeanDaoImpl.getInstance(sContext);
        return logBeanDao.getAll();
    }

    /**
     * 清空数据库
     */
    public static void clearDb() {
        if (sContext == null) {
            throw new RuntimeException("Please init SLog first!");
        }
        final LogBeanDao logBeanDao = LogBeanDaoImpl.getInstance(sContext);
        logBeanDao.deleteAll();
    }

    /**
     * 删除日志文件
     */
    public static void deleteFile() {
        final File file = new File(FILE_PATH);
        if (file.exists()) {
            file.delete();
        }
    }

    public static void changeMode(@LogMode.Check int mode) {
        if (sLogMode != mode) {
            sLogMode = mode;
            destroy();
            sLogStrategy = mode == LogMode.DATABASE ? DatabaseLog.create(sContext) : FileLog.create();
        }
    }

    public static void destroy() {
        flush();
        if (sLogStrategy != null) {
            sLogStrategy.destroy();
            sLogStrategy = null;
        }
    }

    public static void crash(String tag, Throwable t) {
        final LogBean logBean = wrapLogBean(tag, t.toString(), EXCEPTION);
        if (sLogStrategy == null) {
            throw new RuntimeException("Please init SLog first!");
        }
        sLogStrategy.logSync(logBean);
        flush();
    }

    private static void print(String tag, String msg, int priority) {
        synchronized (SLog.class) {
            final LogBean logBean = wrapLogBean(tag, msg, priority);
            if (sLogStrategy == null) {
                throw new RuntimeException("Please init SLog first!");
            }
            sLogStrategy.logAsync(logBean);
        }
    }

}

File: loglib/src/main/java/com/robog/loglib/SLog.kt
package com.robog.loglib

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import com.robog.loglib.FileLog.Companion.FILE_PATH
import com.robog.loglib.Util.wrapLogBean
import com.robog.loglib.database.LogBeanDaoImpl
import java.io.File

/**
 * Created by yuxingdong on 2018/6/29.
 */
@SuppressLint("StaticFieldLeak")
object SLog {

    private const val TAG = "SLog"
    private var logStrategy: LogStrategy? = null
    private var context: Context? = null

    private const val EXCEPTION = -1
    private const val VERBOSE = 2
    private const val DEBUG = 3
    private const val INFO = 4
    private const val WARN = 5
    private const val ERROR = 6

    /**
     * 是否调试模式
     */
    var debug: Boolean = false

    /**
     * 设备标识
     */
    var deviceInfo: String = ""

    /**
     * 数据库检查周期
     */
    var dbCheckInterval: Long = 0

    /**
     * 数据库容量阈值
     */
    var dbThreshold: Int = 0

    /**
     * Log存储模式
     */
    var logMode: LogMode = LogMode.DATABASE

    /**
     * 缓存大小
     */
    var cacheSize: Int = 0


    fun init(context: Context) {
        init(context, LogConfig())
    }

    fun init(context: Context, logConfig: LogConfig) {
        this.context = context.applicationContext

        this.debug = logConfig.debug
        this.deviceInfo = logConfig.deviceInfo
        this.dbCheckInterval = logConfig.dbCheckInterval
        this.dbThreshold = logConfig.dbThreshold
        this.logMode = logConfig.logMode
        this.cacheSize = logConfig.cacheSize

        if (logMode == LogMode.DATABASE) {
            logStrategy = DatabaseLog.create(context.applicationContext)
        } else if (logMode == LogMode.FILE) {
            logStrategy = FileLog.create()
        }
    }

    fun v(tag: String, e: Exception) {
        v(tag, e.toString())
    }

    fun v(msg: String) {
        v(TAG, msg)
    }

    fun v(tag: String, msg: String) {
        if (debug) {
            Log.v(tag, msg)
        }
        print(tag, msg, VERBOSE)
    }

    fun d(tag: String, e: Exception) {
        d(tag, e.toString())
    }

    fun d(msg: String) {
        d(TAG, msg)
    }

    fun d(tag: String, msg: String) {
        if (debug) {
            Log.d(tag, msg)
        }
        print(tag, msg, DEBUG)
    }

    fun i(tag: String, e: Exception) {
        i(tag, e.toString())
    }

    fun i(msg: String) {
        i(TAG, msg)
    }

    fun i(tag: String, msg: String) {
        if (debug) {
            Log.i(tag, msg)
        }
        print(tag, msg, INFO)
    }

    fun w(tag: String, e: Exception) {
        w(tag, e.toString())
    }


    fun w(msg: String) {
        w(TAG, msg)
    }

    fun w(tag: String, msg: String) {
        if (debug) {
            Log.e(tag, msg)
        }
        print(tag, msg, WARN)
    }

    fun e(tag: String, e: Exception) {
        e(tag, e.toString())
    }

    fun e(msg: String) {
        e(TAG, msg)
    }

    fun e(tag: String, msg: String) {
        if (debug) {
            Log.e(tag, msg)
        }
        print(tag, msg, ERROR)
    }

    fun cacheSize(): Int {
        return LogCache.cache.size
    }

    /**
     * 将所有缓存数据推入数据库用于上传
     */
    fun flush() {
        LogCache.get().flush()
    }

    /**
     * 获取所有数据
     */
    fun getAllLog(): List<LogBean> {
        if (context == null) {
            throw RuntimeException("Please init SLog first!")
        }
        val logBeanDao = LogBeanDaoImpl.getInstance(context!!.applicationContext)
        return logBeanDao.getAll()
    }

    /**
     * 清空数据库
     */
    fun clearDb() {
        if (context == null) {
            throw RuntimeException("Please init SLog first!")
        }
        work {
            val logBeanDao = LogBeanDaoImpl.getInstance(context!!.applicationContext)
            logBeanDao.deleteAll()
        }
    }

    /**
     * 删除日志文件
     */
    fun deleteFile() {
        val file = File(FILE_PATH)
        if (file.exists()) {
            work {
                file.delete()
            }
        }
    }

    fun changeMode(mode: LogMode) {
        if (logMode != mode) {
            logMode = mode
            destroy()
            logStrategy =
                    if (mode == LogMode.DATABASE) DatabaseLog.create(context!!.applicationContext)
                    else FileLog.create()
        }
    }

    fun destroy() {
        flush()
        logStrategy?.destroy()
        logStrategy = null
    }

    fun crash(tag: String, t: Throwable?) {
        val logBean = wrapLogBean(tag, t.toString(), EXCEPTION)
        if (logStrategy == null) {
            throw RuntimeException("Please init SLog first!")
        }
        logStrategy!!.logSync(logBean)
        flush()
    }

    private fun print(tag: String, msg: String, priority: Int) {
        synchronized(SLog::class.java) {
            work {
                val logBean = wrapLogBean(tag, msg, priority)
                if (logStrategy == null) {
                    throw RuntimeException("Please init SLog first!")
                }
                logStrategy!!.logAsync(logBean)
            }
        }
    }

}